src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs

```diff
@@ -27,35 +28,76 @@ public abstract class AbstractShardedQueryProcessor<TCommand, TResult, TCombined
-    protected readonly bool _metadataOnly;
-    protected readonly bool _indexEntriesOnly;
-    protected readonly CancellationToken _token;
-
-    protected readonly bool _isMapReduceIndex;
-    protected readonly bool _isAutoMapReduceQuery;
-
     private Dictionary<int, BlittableJsonReaderObject> _queryTemplates;
     private Dictionary<int, TCommand> _commands;
+    private readonly bool _metadataOnly;
+    private readonly bool _indexEntriesOnly;
+    private readonly string _raftUniqueRequestId;
+    private readonly HashSet<int> _filteredShardIndexes;
+
+    protected readonly TransactionOperationContext Context;
+    protected readonly ShardedDatabaseRequestHandler RequestHandler;
+    protected readonly IndexQueryServerSide Query;
+    protected readonly CancellationToken Token;
+    protected readonly bool IsMapReduceIndex;
+    protected readonly bool IsAutoMapReduceQuery;
+    protected readonly long? ExistingResultEtag;
 
     protected AbstractShardedQueryProcessor(TransactionOperationContext context, ShardedDatabaseRequestHandler requestHandler, IndexQueryServerSide query, bool metadataOnly, bool indexEntriesOnly,
-        CancellationToken token)
+        long? existingResultEtag, CancellationToken token)
     {
-        _requestHandler = requestHandler;
-        _query = query;
+        RequestHandler = requestHandler;
+        Query = query;
         _metadataOnly = metadataOnly;
         _indexEntriesOnly = indexEntriesOnly;
-        _token = token;
-        _context = context;
+        Token = token;
+        Context = context;
+        ExistingResultEtag = existingResultEtag;
+
+        IsMapReduceIndex = Query.Metadata.IndexName != null && (RequestHandler.DatabaseContext.Indexes.GetIndex(Query.Metadata.IndexName)?.Type.IsMapReduce() ?? false);
+        IsAutoMapReduceQuery = Query.Metadata.IsDynamic && Query.Metadata.IsGroupBy;
+
+        _raftUniqueRequestId = RequestHandler.GetRaftRequestIdFromQuery() ?? RaftIdGenerator.NewId();
+
+        if (Query.QueryParameters != null && Query.QueryParameters.TryGetMember(Constants.Documents.Querying.Sharding.ShardContextParameterName, out object filter))
+        {
+            // User can define a query parameter ("__shardContext") which is an id or an array
+            // that contains the ids whose shards the query should be limited to.
+            // Advanced: Optimization if user wants to run a query and knows what shards it is on. Such as:
+            // from Orders where State = $state and User = $user where all the orders are on the same share as the user
 
-        _isMapReduceIndex = _query.Metadata.IndexName != null && (_requestHandler.DatabaseContext.Indexes.GetIndex(_query.Metadata.IndexName)?.Type.IsMapReduce() ?? false);
-        _isAutoMapReduceQuery = _query.Metadata.IsDynamic && _query.Metadata.IsGroupBy;
+            _filteredShardIndexes = new HashSet<int>();
+            switch (filter)
+            {
+                case LazyStringValue id:
```

Comment:
can this be a `LazyCompressedStringValue` ?

Reply:
Not very likely but potentially it can. I added it.